presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveBucketing.java

```diff
@@ -90,44 +136,67 @@ private HiveBucketing() {}
 
     public static int getHiveBucket(BucketingVersion bucketingVersion, int bucketCount, List<TypeInfo> types, Page page, int position)
     {
-        return getBucketNumber(getBucketHashCode(bucketingVersion, types, page, position), bucketCount);
+        return getBucketNumber(bucketingVersion.getBucketHashCode(types, page, position), bucketCount);
     }
 
     public static int getHiveBucket(BucketingVersion bucketingVersion, int bucketCount, List<TypeInfo> types, Object[] values)
     {
-        return getBucketNumber(getBucketHashCode(bucketingVersion, types, values), bucketCount);
+        return getBucketNumber(bucketingVersion.getBucketHashCode(types, values), bucketCount);
     }
 
     @VisibleForTesting
-    static int getBucketNumber(int hashCode, int bucketCount)
+    static Optional<Set<Integer>> getHiveBuckets(BucketingVersion bucketingVersion, int bucketCount, List<TypeInfo> types, List<List<NullableValue>> values)
     {
-        return (hashCode & Integer.MAX_VALUE) % bucketCount;
+        long explorationCount;
+        try {
+            explorationCount = values.stream()
+                    .mapToLong(List::size)
+                    .reduce(1, Math::multiplyExact);
+        }
+        catch (ArithmeticException e) {
+            return Optional.empty();
+        }
+        long explorationLimit = Math.max(bucketCount * BUCKETS_EXPLORATION_LIMIT_FACTOR, BUCKETS_EXPLORATION_GRANTED_LIMIT);
```

Comment:
What's the rationale for `BUCKETS_EXPLORATION_LIMIT_FACTOR` vs `BUCKETS_EXPLORATION_GRANTED_LIMIT`? Why not just a single threshold based on the number of possible combinations (i.e., discrete entries)?

Reply:
The idea was to keep the number of combinations checked "reasonably related" to the bucket count.
If the number of combinations highly exceeds the bucket count, then probably all the buckets will be hit. In such case the whole procedure of enumerating the buckets is omitted and no bucket filter is produced (all buckets will be read). I set the threshold to combinations_count = 4*bucketCount.
However, with the above approach, every time we resign from enumerating buckets, we possibly miss the opportunity to filter out some unnecessary buckets. That is why I defined a "GRANTED_LIMIT" = 1000. When the number of combinations is within this limit, all of them will be checked, even if the number of combinations highly exceeds the bucket count.
That is the main idea. The numbers (4, 1000) are arbitrary. They should be backed by some probability calculations and by the knowledge of real use cases.

Reply:
Ok, let's add a short explanation in a comment, as it's not entirely obvious what the motivation is from looking at the code.